src/Agda/Compiler/Rust/Backend.hs

```diff
@@ -77,14 +87,94 @@ compile _ _ _ Defn{..}
   $ getUniqueCompilerPragma "AGDA2RUST" defName >>= \case
       Nothing -> return []
       Just (CompilerPragma _ _) ->
-        return $ prettyShow (qnameName defName) <> " = " <> prettyShow theDef
-
-writeModule :: Options -> ModuleEnv -> IsMain -> TopLevelModuleName -> [CompiledDef]
-            -> TCM ModuleRes
-writeModule opts _ _ m cdefs = do
+        return $ compileDefn defName theDef
+
+compileDefn :: QName
+  -> Defn
+  -> CompiledDef
+compileDefn defName theDef =
+  case theDef of 
+    Datatype{dataCons = fields} ->
+      compileDataType defName fields
+    Function{funCompiled = funDef, funClauses = fc} ->
+      compileFunction defName funDef fc
+    _ ->
+      "UNSUPPORTED " <> showName defName <> " = " <> prettyShow theDef
+
+compileDataType :: QName -> [QName] -> CompiledDef
+compileDataType defName fields = "enum" <> exprSeparator
+  <>  showName defName
+  <> exprSeparator
+  <> bracket (
+    indent
+    <> concat (intersperse ", " (map showName fields)))
+
+compileFunction :: QName
+  -> Maybe CompiledClauses
+  -> [Clause]
+  -> CompiledDef
+compileFunction defName funDef fc = 
+  "pub fn" <> exprSeparator
+    <> showName defName
+    <> "("
+    -- TODO handle multiple function clauses
+    <> compileFunctionArgument (head fc)
+    <> ")" <> exprSeparator <>
+    bracket (
+    -- TODO proper indentation for every line of function body
+    indent
+    <> compileFunctionBody funDef)
+    <> defsSeparator
+
+compileFunctionArgument :: Clause -> CompiledDef
+compileFunctionArgument fc = prettyShow fc
+
+compileFunctionBody :: Maybe CompiledClauses -> CompiledDef
+compileFunctionBody funDef = prettyShow funDef
+
+showName :: QName -> CompiledDef
+showName = prettyShow . qnameName
+
+bracket :: String -> String
+bracket str = "{\n" <> str <> "\n}"
```

Comment:
Function `bracket` and other utilities like: `exprSeparator`, `indent` suggests I am dealing with 2 different responsibilities:
* translate Agda code to Rust and
* pretty printing

Proper solution would be to separate those, by introducing some ADT for Rust and then pretty print. This way consistent indentation should be doable and not that messy.

Switching to this, can wait until I have some basic handling of types and expressions.